### ISSUE TITLE
Array module review followup: Should IRV be renamed to something longer and more descriptive?

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#18088]
This is a followup item for the [Array module review](https://github.com/Cray/chapel-private/issues/2241).

One item we wanted further discussion on is:

whether IRV be renamed to something longer and more descriptive.

This isn't a commonly used function and it would make code more self documenting if we could make the name longer.

**Comment by mppf** [comment]
Should we talk about what it might be renamed to? I'm generally in favor of renaming it but of course it depends on what we were renaming it to.

**Comment by stonea** [comment]
I'm open to suggestions.  To get the conversation started here are a few potential names I can come up with off the top of my head:

1. ZeroValue
2. SparseZeroValue
3. DefaultValue
4. DefaultSparseValue
5. SparseImplicitValue
6. ImplicitlyReplicatedValue
7. ImplicitlyReplicatedValueForSparseArrays
8. ImplicitSparseValue

The language spec already has an argument against calling it a "ZeroValue" so I think we can rule 1 and 2 out.

I like having the word "Sparse" in the name as it gives a clue as to what context this is useful in.  So I would vote against 3 and 6.  Although I think either one is better than IRV.

That leaves 4,5,7 and 8.  7 is very descriptive but seems way too wordy.  I like 4. The term may be a little confusing since for a normal (sparsely populated) sparse array the IRV value itself is not sparsely used (in fact the opposite). The same problem exists for 5, and 8.

**Comment by mppf** [comment]
FWIW I'm pretty happy with DefaultSparseValue / SparseDefaultValue and with SparseImplicitValue / ImplicitSparseValue.

> I like having the word "Sparse" in the name as it gives a clue as to what context this is useful in. So I would vote against 3 and 6. Although I think either one is better than IRV.

I'm not so sure about this. Aren't we talking about a field that can only be used on sparse arrays? So, knowing the type, one would know we're talking about a sparse array? Normally I think we are happy not to name methods with the type e.g. we have `myString.split()` and not `myString.splitString()` and I think the assumption there is that programmers will have an idea what types the are operating on. Why do you feel differently about sparse arrays and the renamed IRV field/method?

**Comment by stonea** [comment]
> I'm not so sure about this. Aren't we talking about a field that can only be used on sparse arrays?

The documentation (https://chapel-lang.org/docs/builtins/ChapelArray.html) lists the function as belonging to the array type.  There isn't a separate SparseArray type right? Rather the behavior of array changes depending on the underlying domain that's associated with it?

It's true the function is only applicable on sparse arrays and will produce an error if someone tries to use on a sparse array. I just tried using it and it looks like we produce this error at compile time:

  `foo.chpl:4: error: only sparse arrays have an IRV`

I guess the confusion I see is that when someone has a rectangular array type and wants to see what they can do they'll go to the documentation and skim the names, or maybe use an IDE feature like intellisense that presents the available methods and they'll see this "IRV" method on their non-sparse array and think that maybe this is relevant to them.

This wouldn't be the case on something like a split() method on a string type as split() is always a valid function to call on something that is a string.

**Comment by lydia-duncan** [comment]
I definitely think we should rename this, and I like DefaultSparseValue / SparseDefaultValue and SparseImplicitValue / ImplicitSparseValue.  I would also be okay with shortening the "Value" portion of the name to "Val" to save some typing.

**Comment by e-kayrakli** [comment]
Nit, but, we are actually considering a lowercase initial for these options, right?

**Comment by bradcray** [comment]
Of these ideas, I like `implicitValue` best.  I don't think we should put `sparse` into the method name for the reason that Michael gives, and w.r.t. Andy's concerns, think we should do a better job of organizing the array documentation so that things that are specific to rectangular / associative / sparse arrays are more clearly organized into sections that make that clear.

I was almost tempted to go with `sparseValue`, less as a way of referring to the "sparse array" nature of the base array, and more because it more directly associated the value's role with its context.  However, by definition, it's really more of the "dense value" in a sparse array (in the sense that there are typically O(n**2) IRVs in a sparse n x n matrix).  So I'm not sure if that would be a step forward.  Brainstorming more, `denseValue` doesn't seem all that clear/useful.  `fillValue` might be an attractive alternative?  It feels somehow descriptive and crisp relative to `implicitValue`.

I'm not crazy about using "default" in the name because I think it has the potential to get mixed up with Chapel's general notion of a default value (e.g., what is the default value used for `var x: t;`?).

[This conversation is also making me remember what we felt some of the benefits of IRV were:  That it's a way of referring to a value that's challenging to describe succinctly.]

Whatever name  we choose, are we imagining a `getSpecialName` / `setSpecialName` interface?  (e.g., `getImplicitValue()`, `setImplicitValue()`?)  Or perhaps a `.specialName` accessor for reading and a `setSpecialName()` method for changing it?

**Comment by mppf** [comment]
Oh, I actually really like `fillValue`. Curious if other people see a major problem with it.

**Comment by e-kayrakli** [comment]
`fillValue` is more of an action, so, it sounds too much like a method name rather than a property/field name to me. I am not too opposed to it, though. I like that it is short.

**Comment by mppf** [comment]
That's true. Maybe we can find a way to make it more noun-y. `implicitValue` doesn't have this problem. Maybe `backgroundValue` or `valueFill` or `fillerValue` ore variations.

**Comment by stonea** [comment]
>  w.r.t. Andy's concerns, think we should do a better job of organizing the array documentation so that things that are specific to rectangular / associative / sparse arrays are more clearly organized into sections that make that clear

Better organization in the documentation would help. Though I think the more we can make the language self documenting the better.

> Whatever name we choose, are we imagining a getSpecialName / setSpecialName interface? (e.g., getImplicitValue(), setImplicitValue()?) Or perhaps a .specialName accessor for reading and a setSpecialName() method for changing it?

Interesting question. Would the answer to it influence what you would like special name to be?  I agree with Engin's concern with the term 'fillvalue' sounding like an action, but if it was called setFillValue\getFillValue I'd be fine with it.

Anyway, given the conversation so far it seems like no one's advocating for 'defaultValue' or any of the suggestions I gave other than implicitValue. So here's what I see on the table so far:

1. implicitValue
2. keeping IRV (or spelling it out as implicitlyReplicatedValue)
3. coming up with something based on the term "fillvalue" (maybe fillerValue, maybe getFillValue\setFillValue)

And then some other questions:
1. Should we try and put the word "sparse" in the name.
2. Short we shorten value to val
3. Should we have get\set methods.

I think my vote would be for using 'implicitValue' if we decide to not have get\set methods and using 'fillValue' if we did.

**Comment by bradcray** [comment]
> Though I think the more we can make the language self documenting the better.

I don't disagree with this, though there's a huge challenge to doing that in the array case given that the `_array` type on which all these methods are defined is not user-facing / officially part of the language.  At times we've discussed making it more first-class (in which case using a partial instantiation of that type could be a way of saying "this method is for sparse arrays), but we haven't ever put much effort into deciding whether to do it or making that happen.  That's what led me to suggest using sections in the docs, at least for the time being.

> Would the answer to it influence what you would like special name to be? I agree with Engin's concern with the term 'fillvalue' sounding like an action, but if it was called setFillValue\getFillValue I'd be fine with it.

Yeah, that was my thinking/concern as well.  If I were championing `fillValue`, I might support the getter as a paren-less property method:  `myArray.fillValue` which makes it read slightly more like a noun by not looking like an action but a property.  And then something like `myArray.setFillValue(val: eltType)` to set it.  To some extent, I think this question is one that's been lurking just beneath the surface in some of these module reviews, which is how we want to deal with getter / setter like things in the library, where some possible options are:

* symmetric get()/set() routines
* paren-less get-routines / paren-ful set routines
* paren-less get and set routines (though I don't believe that will work in this case, at least easily, due to the need to do more than a simple assignment for distributed sparse arrays).

> I think my vote would be for using 'implicitValue' if we decide to not have get\set methods

What would this look like?  A `ref` and non-`ref` overload of a paren-less `.implicitValue` method?  (my bullet 3 above).  To put some more flesh on my concern, I think this issue captures it:  https://github.com/chapel-lang/chapel/issues/16889  (and this comment specifically:  https://github.com/chapel-lang/chapel/issues/16889#issuecomment-754190943).

**Comment by bradcray** [comment]
@lydia-duncan : I think we've agreed that we're not committing to stabilizing spare arrays for 2.0, so it seems like this shouldn't have the Chapel 2.0 label?

**Comment by lydia-duncan** [comment]
I think it needs to be marked unstable if it isn't already but am happy to remove the 2.0 label when that happens

**Comment by bradcray** [comment]
I disagree with that.  We've discussed making any program that refers to `sparse` generate an unstable warning, and that's a reasonable thing to desire in the 2.0 timeframe.  But if it's a requirement for 2.0 rather than just "nice to have", an issue (or PR) should be opened for it and the 2.0 label put on that issue rather than using this one as a stand-in for that desire.

Put another way, I don't think having every operation one might do on a sparse domain or array generate its own unstable warning is attractive or desirable, either from the implementer's perspective or the user's.  Rather, I'd expect that any program referring to `sparse` would generate a single blanket "sparse is not considered stable" type of warning no matter how much or little it does with sparse domains and arrays.

**Comment by lydia-duncan** [comment]
Okay, I buy that

**Comment by bradcray** [comment]
Last night, I was wondering whether `.IRV` should just be `.default` and whether we should support that for dense arrays as well (e.g., when resized, what should the default value for new elements be?).  I'm noting it here because it looks like I was opposed to it in the past.  I think the things that caused me to flip were the potential to apply it to sparse as well as dense and the brevity / single-wordness.  But I didn't re-read this whole issue, so just wanted to log this while it occurred to me.
